parser/types/field_type.go

```diff
@@ -38,26 +39,106 @@ var (
 
 // FieldType records field type information.
 type FieldType struct {
-	Tp      byte
-	Flag    uint
-	Flen    int
-	Decimal int
-	Charset string
-	Collate string
-	// Elems is the element list for enum and set type.
-	Elems []string
+	tp      byte
+	flag    uint
+	flen    int
+	decimal int
+	charset string
+	collate string
+	// elems is the element list for enum and set type.
+	elems []string
 }
 
 // NewFieldType returns a FieldType,
 // with a type and other information about field type.
 func NewFieldType(tp byte) *FieldType {
 	return &FieldType{
-		Tp:      tp,
-		Flen:    UnspecifiedLength,
-		Decimal: UnspecifiedLength,
+		tp:      tp,
+		flen:    UnspecifiedLength,
+		decimal: UnspecifiedLength,
 	}
 }
 
+func (ft *FieldType) GetType() byte {
+	return ft.tp
+}
+
+func (ft *FieldType) GetFlag() uint {
+	return ft.flag
+}
+
+func (ft *FieldType) GetFlen() int {
+	return ft.flen
+}
+
+func (ft *FieldType) GetDecimal() int {
+	return ft.decimal
+}
+
+func (ft *FieldType) GetCharset() string {
+	return ft.charset
+}
+
+func (ft *FieldType) GetCollate() string {
+	return ft.collate
+}
+
+func (ft *FieldType) GetElems() []string {
+	return ft.elems
+}
+
+func (ft *FieldType) SetType(tp byte) {
+	ft.tp = tp
+}
+
+func (ft *FieldType) SetFlag(flag uint) {
+	ft.flag = flag
+}
+
+func (ft *FieldType) AddFlag(flag uint) {
+	ft.flag |= flag
+}
+
+func (ft *FieldType) AndFlag(flag uint) {
+	ft.flag &= flag
+}
+
+func (ft *FieldType) ToggleFlag(flag uint) {
+	ft.flag ^= flag
+}
+
+func (ft *FieldType) DelFlag(flag uint) {
+	ft.flag &= ^flag
+}
+
+func (ft *FieldType) SetFlen(flen int) {
+	ft.flen = flen
+}
+
+func (ft *FieldType) SetDecimal(decimal int) {
+	ft.decimal = decimal
+}
+
+func (ft *FieldType) SetCharset(charset string) {
+	ft.charset = charset
+}
+
+func (ft *FieldType) SetCollate(collate string) {
+	ft.collate = collate
+}
+
+func (ft *FieldType) SetElems(elems []string) {
```

Comment:
Should we set `element.clone()` to `ft.elems[idx]`? I think it is unsafe to directly assign the value.

Reply:
That is a logic changer. This PR so far, try to avoid logic changes as much as possible. Only setters/getters, field type builder are added to make things consistent. I think it is too early to adjust COW or any other read/write isolation mechanism.

BTW, my preference would be a read only `FieldType` by default, and explicitly copy `MutableFieldType`.